File: 백준/Gold/1717. 집합의 표현/집합의 표현.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	public class UnionFind {
		int[] parent;
		int count;

		public UnionFind(int count) {
			this.count = count;
			parent = new int[this.count];
		}

		public void makeSet() {
			for (int i = 0; i < count; i++) {
				parent[i] = -1;
			}
		}

		private int find(int num) {
			if (parent[num] < 0) {
				return num;
			}

			return parent[num] = find(parent[num]);
		}

		private void merge(int num1, int num2) {
			int rootA = find(num1);
			int rootB = find(num2);

			if (rootA == rootB) {
				return;
			}

			int temp = parent[rootA] + parent[rootB];

			if (parent[rootA] > parent[rootB]) {
				parent[rootA] = rootB;
				parent[rootB] = temp;
			} else {
				parent[rootB] = rootA;
				parent[rootA] = temp;
			}
		}
	}

	public void solution() throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		StringBuilder sb = new StringBuilder();

		int N = Integer.valueOf(st.nextToken());
		int M = Integer.valueOf(st.nextToken());

		UnionFind uf = new UnionFind(N + 1);
		uf.makeSet();

		for (int i = 0; i < M; i++) {
			st = new StringTokenizer(br.readLine());

			int command = Integer.valueOf(st.nextToken());
			int a = Integer.valueOf(st.nextToken());
			int b = Integer.valueOf(st.nextToken());

			if (command == 0) {
				uf.merge(a, b);
			}

			if (command == 1) {
				int rootA = uf.find(a);
				int rootB = uf.find(b);

				if (rootA == rootB) {
					sb.append("YES")
						.append("\n");
				} else {
					sb.append("NO")
						.append("\n");
				}
			}
		}
		System.out.println(sb);
	}

	public static void main(String[] args) throws IOException {
		new Main().solution();
	}
}

File: 백준/Gold/1717. 집합의 표현/집합의 표현.js
const fs = require('fs');
// const input = fs.readFileSync('input.txt').toString().trim().split('\n');
const input = fs.readFileSync(0).toString().trim().split('\n');
class UnionFind {
  constructor(count) {
    this.count = count;
    this.parent = new Array(count).fill(-1);
  }

  find(num) {
    if (this.parent[num] < 0) return num;
    return this.parent[num] = this.find(this.parent[num]);
  }

  merge(num1, num2) {
    const rootA = this.find(num1);
    const rootB = this.find(num2);
    if (rootA === rootB) return;

    const temp = this.parent[rootA] + this.parent[rootB];
    if (this.parent[rootA] > this.parent[rootB]) {
      this.parent[rootA] = rootB;
      this.parent[rootB] = temp;
    } else {
      this.parent[rootB] = rootA;
      this.parent[rootA] = temp;
    }
  }
}

function solution(input) {
  const [N, M] = input[0].split(' ').map(Number);
  const uf = new UnionFind(N + 1);
  const output = [];

  for (let i = 1; i <= M; i++) {
    const [command, a, b] = input[i].split(' ').map(Number);
    if (command === 0) {
      uf.merge(a, b);
    } else if (command === 1) {
      output.push(uf.find(a) === uf.find(b) ? "YES" : "NO");
    }
  }

  console.log(output.join('\n'));
}

solution(input);
